### ISSUE TITLE
CentroidNet - A Deep Neural Network for Joint Object Localization and Counting

### ISSUE THREAD

**Issue by junxnone** [junxnone/tio#855]
# Reference
- [CentroidNet: A Deep Neural Network for Joint Object Localization and Counting](https://github.com/junxnone/tech-io/files/5281457/CentroidNet.A.Deep.Neural.Network.for.Joint.pdf)
- [Code - pytorch](https://github.com/kdijkstra13/OpenCentroidNet)

# Brief

## Preprocess
- 靠近边界的 BBox 被移除
- `Target Image` - `Label` 
  - `2 + 1 + N` 通道 numpy array
    - **2** - Vector/指向最近目标/类似于投票
    - **1** - 背景
    - **N** - 目标类别数  
  - [x_centroid_vector, y_centroid_vector, bbox_ellipse[0:n_classs]]
    - **Channel 0** - 每个像素质心向量y分量
    - **Channel 1** - 每个像素质心向量x分量
    - **Channel 2** - 背景区域设置为`1`，目标内接圆区域设置为`0`
    - **Channel 2+N** - bbox取内接椭圆内部像素，相应类别`N`设置为 1

---
Input Image | Vector-X | Vector-Y | Background | Object
-- | -- | -- | -- | --
![image](https://user-images.githubusercontent.com/2216970/99473870-729a3f80-2986-11eb-810a-c0ec0962182f.png) | ![image](https://user-images.githubusercontent.com/2216970/99473895-80e85b80-2986-11eb-909a-d13c44576416.png) | ![image](https://user-images.githubusercontent.com/2216970/99473907-880f6980-2986-11eb-9e20-6316db46f9d5.png) | ![image](https://user-images.githubusercontent.com/2216970/99473921-8fcf0e00-2986-11eb-86e9-efd7a18ee852.png) | ![image](https://user-images.githubusercontent.com/2216970/99473931-95c4ef00-2986-11eb-8249-1c63f4b0a69f.png)
![image](https://user-images.githubusercontent.com/2216970/99891792-c23f7a80-2ca8-11eb-99cb-daac7e99cbbd.png) | ![image](https://user-images.githubusercontent.com/2216970/99891794-cb304c00-2ca8-11eb-9ec7-258ef8ec75af.png) | ![image](https://user-images.githubusercontent.com/2216970/99891800-dc795880-2ca8-11eb-80b2-121bf97b8c4f.png) | ![image](https://user-images.githubusercontent.com/2216970/99891802-e4d19380-2ca8-11eb-8b6f-132ac54e7d9d.png) | ![image](https://user-images.githubusercontent.com/2216970/99891897-c91abd00-2ca9-11eb-81b1-9b64bd2f7d60.png)

> mask 理论上是个椭圆, 使用 `skimage.draw.ellipse()` 绘制，`Input - 1` 中 `Object` 太小，成了矩形

## Vector/Distance 计算
- 计算 `Vector`  
  - 目标中心坐标和当前像素坐标的差
- 计算 `Distance Map`
  - Shape = (N, W, H)
  - N 为 `Object` 数量
  - 当前像素到Object 中心位置的**欧氏距离**
- 对于 (N, W, H) 的 N 个 Object 的 distance map 合并，取distance最小值 作为 Index，取相应的 (Vector-X, Vector-Y)
- 大于`Max Distance` 的 (Vector-X, Vector-Y) 设置为 (0, 0)

---
![image](https://user-images.githubusercontent.com/2216970/99498761-eb16f580-29b2-11eb-8279-9e28b10804cd.png) | ![image](https://user-images.githubusercontent.com/2216970/99498980-2e716400-29b3-11eb-8262-92062fd9afcf.png)
-- | --

Vector-X | Vector-Y
-- |--
![image](https://user-images.githubusercontent.com/2216970/99503442-2b797200-29b9-11eb-9d3d-1ebfd4370056.png) | ![image](https://user-images.githubusercontent.com/2216970/99503478-36340700-29b9-11eb-89f5-e93b5fe422a3.png)

## Network Arch
- Unet 架构 `Down-5/Up-4`

Downscaling Pathway | Upscaling Pathway  
-- | -- 
![image](https://user-images.githubusercontent.com/2216970/99479052-76cb5a80-2990-11eb-80dd-65523939289b.png) | ![image](https://user-images.githubusercontent.com/2216970/99479076-7d59d200-2990-11eb-8059-a06aba9684f6.png)

Tensorboard | Pathway
-- | --
![image](https://user-images.githubusercontent.com/2216970/99476107-e63e4b80-298a-11eb-9580-7d2efbe909f7.png)| ![image](https://user-images.githubusercontent.com/2216970/99476031-bf801500-298a-11eb-835b-4cbd9774028a.png)

## Voting
![image](https://user-images.githubusercontent.com/2216970/94532083-fcc00480-026f-11eb-973e-37a0bcbba199.png)

**Comment by junxnone** [comment]
#845
